fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java

```diff
@@ -1268,18 +1273,39 @@ private Map<Integer, Map<String, ColumnRefOperator>> getRelationIdToColumns(Colu
         return result;
     }
 
+    private boolean isCompensationComplete(ScalarOperator rewritten,
+                                           ColumnRefSet joinOnPredicateColumnRefSet) {
+        // step1: Check compensation rewritten is ok.
+        if (rewritten == null) {
+            return false;
+        }
+
+        // step2: Check whether all necessary join on predicates are all rewritten.
+        {
+            if (joinOnPredicateColumnRefSet == null || joinOnPredicateColumnRefSet.isEmpty()) {
+                return true;
+            }
+            if (!isAllExprReplaced(rewritten, joinOnPredicateColumnRefSet)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
     // when isQueryAgainstView is true, get compensation predicates of query against view
     // or get compensation predicates of view against query
-    private PredicateSplit getCompensationPredicates(RewriteContext rewriteContext, boolean isQueryAgainstView) {
+    private PredicateSplit getCompensationPredicates(RewriteContext rewriteContext,
+                                                     ColumnRefSet joinOnPredicateColumnRefSet,
+                                                     boolean isQueryAgainstView) {
         // 1. equality join subsumption test
         EquivalenceClasses sourceEquivalenceClasses = isQueryAgainstView ?
                 rewriteContext.getQueryEquivalenceClasses() : rewriteContext.getQueryBasedViewEquivalenceClasses();
         EquivalenceClasses targetEquivalenceClasses = isQueryAgainstView ?
                 rewriteContext.getQueryBasedViewEquivalenceClasses() : rewriteContext.getQueryEquivalenceClasses();
         final ScalarOperator compensationEqualPredicate =
                 getCompensationEqualPredicate(sourceEquivalenceClasses, targetEquivalenceClasses);
-        if (compensationEqualPredicate == null) {
-            // means source equal predicates cannot be rewritten by target
+        if (!isCompensationComplete(compensationEqualPredicate, joinOnPredicateColumnRefSet)) {
```

Comment:
if query is:
select count(1)
    //          from customer left outer join orders on c_custkey = o_custkey
    //          where  c_custkey = 1
can this query pass this check?

Reply:
Add a unit test.